src/tests/Loader/classloader/StaticVirtualMethods/DiamondShape/svm_diamondshape.il

```diff
@@ -0,0 +1,857 @@
+
+  // --- The following custom attribute is added automatically, do not uncomment -------
+  //  .custom instance void [mscorlib]System.Diagnostics.DebuggableAttribute::.ctor(valuetype [mscorlib]System.Diagnostics.DebuggableAttribute/DebuggingModes) = ( 01 00 07 01 00 00 00 00 ) 
+
+  .hash algorithm 0x00008004
+  .ver 0:0:0:0
+}
+// MVID: {25B467F3-D284-4DB6-BAD4-C7EAC75CF064}
+.imagebase 0x00400000
+.file alignment 0x00000200
+.stackreserve 0x00100000
+.subsystem 0x0003       // WINDOWS_CUI
+.corflags 0x00000001    //  ILONLY
+// Image base: 0x010B0000
+
+
+// =============== CLASS MEMBERS DECLARATION ===================
+
+.class interface private abstract auto ansi IFoo
+{
+  .method public hidebysig newslot virtual
+          static int32  Foo(int32 a) cil managed
+  {
+    // Code size       7 (0x7)
+    .maxstack  1
+    .locals init (int32 V_0)
+    IL_0000:  nop
+    IL_0001:  ldarg.0
+    IL_0002:  stloc.0
+    IL_0003:  br.s       IL_0005
+
+    IL_0005:  ldloc.0
+    IL_0006:  ret                 
+  } // end of method IFoo::Foo
+
+} // end of class IFoo
+
+.class interface private abstract auto ansi IFoo2
+       implements IFoo
+{
+  .method public hidebysig final 
+          static int32  Foo(int32 a) cil managed
+  {
+    .override method int32 IFoo::Foo(int32)
+    // Code size       20 (0x14)
+    .maxstack  2
+    .locals init (int32 V_0)
+    IL_0000:  nop
+    IL_0001:  ldstr      "At IFoo2.Foo"
+    IL_0006:  call       void [mscorlib]System.Console::WriteLine(string)
+    IL_000b:  nop
+    IL_000c:  ldarg.0
+    IL_000d:  ldc.i4.1
+    IL_000e:  add
+    IL_000f:  stloc.0
+    IL_0010:  br.s       IL_0012
+
+    IL_0012:  ldloc.0
+    IL_0013:  ret
+  } // end of method IFoo2::IFoo.Foo
+
+} // end of class IFoo2
+
+.class interface private abstract auto ansi IFooEx
+       implements IFoo
+{
+  .method public hidebysig final 
+          static int32  IFoo.Foo(int32 a) cil managed
+  {
+    .override method int32 IFoo::Foo(int32)
+    // Code size       20 (0x14)
+    .maxstack  2
+    .locals init (int32 V_0)
+    IL_0000:  nop
+    IL_0001:  ldstr      "At IFooEx.Foo"
+    IL_0006:  call       void [mscorlib]System.Console::WriteLine(string)
+    IL_000b:  nop
+    IL_000c:  ldarg.0
+    IL_000d:  ldc.i4.2
+    IL_000e:  add
+    IL_000f:  stloc.0
+    IL_0010:  br.s       IL_0012
+
+    IL_0012:  ldloc.0
+    IL_0013:  ret
+  } // end of method IFooEx::IFoo.Foo
+} // end of class IFooEx
+
+.class private auto ansi beforefieldinit FooClass
+       extends [mscorlib]System.Object
+       implements IFoo2,
+                  IFoo,
+                  IFooEx
+{
+} // end of class FooClass
+
+.class interface private abstract auto ansi I1
+{
+  .method public hidebysig newslot abstract virtual 
+          static int32  Func(int32 a) cil managed
+  {
+  } // end of method I1::Func
+
+} // end of class I1
+
+.class interface private abstract auto ansi I2
+       implements I1
+{
+  .method private hidebysig final 
```

Comment:
<pre>
Unhandled exception. System.TypeLoadException: Method implementation on an interface 'IFoo2' from assembly 'svm_diamondshape, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' must be a final method.
</pre>

Am I right to understand this is an artifact of the pre-existing limitation that there can always be just one "actual" override? Should I strive to remove this limitation as it seems applicable no more?

Reply:
This seems odd. These are static methods, and I don't believe final makes sense on static methods. I don't believe we need the static flag on implementation on the class either.

Reply:
> These are static methods, and I don't believe final makes sense on static methods.

I agree with this.

> I don't believe we need the static flag on implementation on the class either.

I am not sure I agree with this. The implemented method is static, so I think the implementing method should be marked static too. BTW, we do mark implementations in classes like that. Is there a reason. Here is what the spec says
```
II.23.1.10 Flags for methods [MethodAttributes]
Static 0x0010 Defined on type, else per instance  
```

Reply:
BTW, I think `final` should be used for re-abstraction. Such instance methods are marked "abstract virtual final" and static methods will be "static abstract virtual final". Do we have any examples for re-abstraction for static methods?

Reply:
@AlekseyTs - am I right to understand that by re-abstraction you mean something like a "newslot" in the sense of introducing a place in the type hierarchy where you'd suppress the SVM implementations defined "at a lower level" and require new ones to be supplied higher up the type hierarchy? If my interpretation is correct, I don't recall any targeted testing for this in our original SVM implementation but it shouldn't be rocket science to expand on the SVM diamond pattern test to include this case. I apologize if I misunderstood you, please clarify in such case.

Reply:
I'll respond with a code example because we are using many "previously undefined terms":
```
interface I1
{
    static abstract void M();
}

interface I2 : I1
{
    static void I1.M(){}
}

interface I3 : I2
{
    static abstract void I1.M(); // re-abstraction
}

class C2 : I2 {} // All is good

class C3: I3 {} // error CS0535: 'C3' does not implement interface member 'I1.M()'
```

It works the same way for instance methods today in the compiler. And it is supposed to work this way for instance methods in the runtime too.

Reply:
Thanks Aleksey, I believe your code example matches what I was trying to describe. I assume that the "diamond pattern" here may mean that in one edge of the type hierarchy graph the SVM has been "re-abstracted" while in another "the original SVM is implemented" thus forming another type of resolution ambiguity. We should certainly include some of these corner cases in unit testing for the feature.

Reply:
>  I believe your code example matches what I was trying to describe.

Great! It still looks like we need to get to a firm agreement around flags used in metadata for method declarations in each scenario.